src/app/InteractionModelEngine.cpp

```diff
@@ -1434,5 +1435,41 @@ size_t InteractionModelEngine::GetNumDirtySubscriptions() const
     return numDirtySubscriptions;
 }
 
+void InteractionModelEngine::OnFabricRemoved(const FabricTable & fabricTable, FabricIndex fabricIndex)
+{
+    mReadHandlers.ForEachActiveObject([fabricIndex](ReadHandler * handler) {
+        if (handler->GetAccessingFabricIndex() == fabricIndex)
+        {
+            ChipLogProgress(InteractionModel, "Fabric removed, deleting obsolete read handler with FabricIndex: %u", fabricIndex);
+            handler->Close();
+        }
+
+        return Loop::Continue;
+    });
+
+    for (auto * readClient = mpActiveReadClientList; readClient != nullptr; readClient = readClient->GetNextClient())
+    {
+        if (readClient->GetFabricIndex() == fabricIndex)
+        {
+            ChipLogProgress(InteractionModel, "Fabric removed, deleting obsolete read client with FabricIndex: %u", fabricIndex);
+            readClient->Close(CHIP_NO_ERROR, false);
+            RemoveReadClient(readClient);
```

Comment:
Why do we need to do this manual removal at all?

Reply:
since the pointer removal is happening inside read client destructor.

ReadClient::~ReadClient()
{
    if (IsSubscriptionType())
    {
        CancelLivenessCheckTimer();
        CancelResubscribeTimer();

        //
        // Only remove ourselves from the engine's tracker list if we still continue to have a valid pointer to it.
        // This won't be the case if the engine shut down before this destructor was called (in which case, mpImEngine
        // will point to null)
        //
        if (mpImEngine)
        {
            mpImEngine->RemoveReadClient(this);
        }
    }
}

Reply:
> since the pointer removal is happening inside read client destructor.

OK, but so what?  What's special about this Close() call compared to other Close() calls that don't do the manual removal?

Reply:
per chat with boris, we should do the same thing here no matter whether the Close is coming from this code or from an error on the wire: we should either remove in both cases or not remove in both cases. 
I just remove the manual removal